src/Arcus.WebApi.Security/Authentication/SharedAccessKey/SharedAccessKeyAuthenticationFilter.cs

```diff
@@ -51,59 +57,125 @@ public async Task OnAuthorizationAsync(AuthorizationFilterContext context)
-                {
-                    if (context.HttpContext.Request.Query[_queryParameterName] != foundSecret)
-                    {
-                        context.Result = new UnauthorizedResult();
-                    }
-                }
-
+                ValidateSharedAccessKeyInRequestHeader(context, foundSecret, logger);
+                ValidateSharedAccessKeyInQueryParameter(context, foundSecret, logger);
             }
         }
 
         private async Task<string> GetAuthorizationSecretAsync(AuthorizationFilterContext context)
         {
             ISecretProvider userDefinedSecretProvider =
-                                context.HttpContext.RequestServices.GetService<ICachedSecretProvider>()
-                                ?? context.HttpContext.RequestServices.GetService<ISecretProvider>();
+                context.HttpContext.RequestServices.GetService<ICachedSecretProvider>()
+                ?? context.HttpContext.RequestServices.GetService<ISecretProvider>();
 
-            if (userDefinedSecretProvider == null)
+            if (userDefinedSecretProvider is null)
             {
                 throw new KeyNotFoundException(
                     $"No configured {nameof(ICachedSecretProvider)} or {nameof(ISecretProvider)} implementation found in the request service container. "
                     + "Please configure such an implementation (ex. in the Startup) of your application");
             }
 
-            string foundSecret = await userDefinedSecretProvider.GetRawSecretAsync(_secretName);
-            if (foundSecret == null)
+            Task<string> rawSecretAsync = userDefinedSecretProvider.GetRawSecretAsync(_secretName);
+            if (rawSecretAsync is null)
+            {
+                throw new InvalidOperationException(
+                    $"Configured {nameof(ISecretProvider)} is not implemented correctly as it returns 'null' for a {nameof(Task)} value when calling {nameof(ISecretProvider.GetRawSecretAsync)}");
+            }
+
+            string foundSecret = await rawSecretAsync;
+            if (foundSecret is null)
             {
                 throw new SecretNotFoundException(_secretName);
             }
 
             return foundSecret;
         }
+
+        private void ValidateSharedAccessKeyInRequestHeader(AuthorizationFilterContext context, string foundSecret, ILogger logger)
+        {
+            if (String.IsNullOrWhiteSpace(_headerName))
+            {
+                return;
+            }
+
+            if (context.HttpContext.Request.Headers.TryGetValue(_headerName, out StringValues requestSecretHeaders))
+            {
+                if (requestSecretHeaders.Any(headerValue => headerValue != foundSecret))
+                {
+                    LogSecurityEvent(logger, SecurityResult.Failure, $"Shared access key in request header '{_headerName}' doesn't match expected access key");
```

Comment:
Do we really need 2 log entries?

Reply:
Hmm, I know I was thinking about that too. I want to log this as a security event, but I also as an error... Maybe we should in our Observability package also have the opportiunity to have success/failure/neither differences.

Reply:
Included the log level in the security event so we can use a single log event.